src/span_ext.rs

```
@@ -214,18 +215,36 @@ pub trait OpenTelemetrySpanExt {
 }
 
 impl OpenTelemetrySpanExt for tracing::Span {
+    ///
+    /// Allows us to set the parent context of this span. This method exists primarily to allow
+    /// us to pull in distributed_ incoming context - e.g. span IDs, etc - that have been read
+    /// into an existing context.
+    ///
+    /// A span's parent should only be set _once_, for the purpose described above.
+    /// Additionally, once a span has been fully built - and the SpanBuilder has been consumed -
+    /// the parent _cannot_ be mutated.
+    ///
     fn set_parent(&self, cx: Context) {
```

I think this should return a `Result` to signal whether this actually did something or not. I get that it's technically a breaking change and you tried to keep the api identical, but unless someone actually did `let () = span.set_parent(...);` which I highly doubt, or someone implemented this trait for their own type (which I doubt and they probably should not have done so either), users will get at most a warning about an unused `Result`.

Other methods that behave differently based on whether the context was already started or not should do the same thing, but I don't think anything else has this kind of behavior.

---

Sure, that's a valid point.

---

While looking at this I realized that not only this one but any of the setters in the `OpenTelemetrySpanExt` can _fail_ silently if there is no `OtelData` associated with the `Span`. Adding a `Result` here will change a lot of code, and I'm not sure if there is any benefit to this since there is nothing the end user can do about it (except maybe log something).

---

The end user can at least unwrap on it in their prototype phase. The alternative being set all of tracing and propagation up, export their traces and then manually check that all their traces have correct parents.

Ignoring the `Result` is easy enough for users that I think that giving them the option to at least log the issue (and the parent context they tried to use) even if they can't handle the failure properly is worth it. And I do believe it's too easy to misuse this API that I'd rather have everyone explicitly ignore the `Result`s, because I believe that this will happen.

I guess we can ignore the other methods for now then. However, I think that here it would be worth it to cover all cases because if setting the parent fails for any reason, the whole trace will be broken as even the trace ID will not be propagated anywhere else and every child of this broken span will be a root span of its own new trace. I get there probably isn't much the code can do if that happens in production, but I still believe we should not just silently ignore it.

---

I've updated `set_parent` to return a result.